src/alacris_lib/state_update.ml

```diff
@@ -87,34 +55,75 @@ let post_state_update_nocheck : operator:Address.t -> operator_revision:Revision
 let rec inner_state_update_periodic_loop : Address.t -> unit Lwt_exn.t =
   fun operator ->
   let open Lwt_exn in
-  retrieve_validated_rev_digest ()
-  >>= fun (operator_revision, operator_digest) -> post_state_update ~operator ~operator_revision ~operator_digest
+  retrieve_validated_state_update ()
+  >>= fun confirmed_state_update -> post_state_update ~operator ~confirmed_state_update
   >>= fun () -> sleep_delay_exn Side_chain_server_config.state_update_period_in_seconds_f
   >>= fun () -> inner_state_update_periodic_loop operator
 
-let rec inner_state_update_nocheck_periodic_loop : Address.t -> unit Lwt_exn.t =
-  fun operator ->
-  let open Lwt_exn in
-  retrieve_validated_rev_digest ()
-  >>= fun (operator_revision, operator_digest) -> post_state_update_nocheck ~operator ~operator_revision ~operator_digest
-  >>= fun () -> sleep_delay_exn Side_chain_server_config.state_update_period_in_seconds_f
-  >>= fun () -> inner_state_update_nocheck_periodic_loop operator
-
 let start_state_update_periodic_daemon address =
-  register_keypair "alice" Signing.Test.alice_keys;
-  register_keypair "bob" Signing.Test.bob_keys;
+  register_keypair "heckle" Signing.Test.heckle_keys;
+  register_keypair "jeckle" Signing.Test.jeckle_keys;
   if state_update_log then
     Logging.log "Beginning of start_state_update_periodic_operator wait=%f" Side_chain_server_config.state_update_period_in_seconds_f;
   Lwt.async (fun () -> inner_state_update_periodic_loop address);
   Lwt_exn.return ()
 
-let start_state_update_nocheck_periodic_operator address =
-  if state_update_log then
-    Logging.log "Beginning of start_state_update_nocheck_periodic_operator wait=%f" Side_chain_server_config.state_update_period_in_seconds_f;
-  Lwt.async (fun () -> inner_state_update_nocheck_periodic_loop address);
-  Lwt_exn.return ()
-
 
+module Test = struct
+
+  let post_state_update_for_test : operator:Address.t -> confirmed_state_update:StateUpdate.t -> unit Lwt_exn.t =
+    fun ~operator ~confirmed_state_update ->
+    let open Lwt_exn in
+    if state_update_log then
+      Logging.log "post_state_update beginning";
+    let null_oper = ref false in
```

Comment:
Instead of a ref, move the comparison + explanation to a separate, pure, function that returns an option for an explanation of difference, or a boolean with logging as side-effect.

Reply:
Also, maybe have a function decode a maybe-zero `digest` into a `digest option`, then directly compare the options?

Reply:
Instead of having an ad-hoc thing for state updates, I would much rather a regular non-test daemon, and have the entire keepalive be made part of `Ethereum_user` or `Ethereum_transaction` while confirming a transaction.

Maybe add a TODO comment to that effect?

Reply:
TODO put.